cupy/_core/dlpack.pyx

```diff
@@ -18,140 +13,244 @@ import warnings
     stdlib.free(tensor)
 
 
-# The name of this function is following the framework integration guide of
-# TensorComprehensions.
-cpdef object toDlpack(_ndarray_base array) except +:
-    cdef DLManagedTensor* dlm_tensor = \
-        <DLManagedTensor*>stdlib.malloc(sizeof(DLManagedTensor))
-
-    cdef size_t ndim = array._shape.size()
-    cdef DLTensor* dl_tensor = &dlm_tensor.dl_tensor
-    cdef intptr_t data_ptr = array.data.ptr
-    dl_tensor.data = <void*>data_ptr
-    dl_tensor.ndim = ndim
+cdef void deleter_ver(DLManagedTensorVersioned* tensor) with gil:
+    # Delete fully initialized DLManagedTensorVersioned
+    stdlib.free(tensor.dl_tensor.shape)
+    cpython.Py_DECREF(<_ndarray_base>tensor.manager_ctx)
+    tensor.manager_ctx = NULL
+    stdlib.free(tensor)
 
-    cdef int64_t* shape_strides = \
-        <int64_t*>stdlib.malloc(ndim * sizeof(int64_t) * 2)
-    for n in range(ndim):
-        shape_strides[n] = array._shape[n]
-    dl_tensor.shape = shape_strides
-    for n in range(ndim):
-        shape_strides[n + ndim] = array._strides[n] // array.dtype.itemsize
 
-    dl_tensor.strides = shape_strides + ndim
-    dl_tensor.byte_offset = 0
+cdef uint8_t get_dlpack_dtype_code(dtype):
+    """Convert NumPy/CuPy to dlpack dtype (kind, without bitsize).
+    """
+    kind = dtype.kind
+
+    if kind == 'u':
+        return <uint8_t>kDLUInt
+    elif kind == 'i':
+        return <uint8_t>kDLInt
+    elif kind == 'f':
+        return <uint8_t>kDLFloat
+    elif kind == 'c':
+        return <uint8_t>kDLComplex
+    elif kind == 'b':
+        return <uint8_t>kDLBool
+    else:
+        raise BufferError('dtype is not supported for dlpack export')
 
-    cdef DLDevice* device = &dl_tensor.device
+
+cdef DLDevice get_dlpack_device(_ndarray_base array):
+    cdef DLDevice device
     cdef bint is_managed
-    cdef int dev_id = array.data.device_id
+
+    device.device_id = array.data.device_id
+
     if not runtime._is_hip_environment:
-        attrs = runtime.pointerGetAttributes(data_ptr)
+        attrs = runtime.pointerGetAttributes(array.data.ptr)
         is_managed = (attrs.type == runtime.memoryTypeManaged)
         if is_managed:
             device.device_type = kDLCUDAManaged
-            dev_id = 0  # make it accessible on CPU too
         else:
             device.device_type = kDLCUDA
     else:
         device.device_type = kDLROCM
-    device.device_id = dev_id
-
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    if array.dtype.kind == 'u':
-        dtype.code = <uint8_t>kDLUInt
-    elif array.dtype.kind == 'i':
-        dtype.code = <uint8_t>kDLInt
-    elif array.dtype.kind == 'f':
-        dtype.code = <uint8_t>kDLFloat
-    elif array.dtype.kind == 'c':
-        dtype.code = <uint8_t>kDLComplex
-    elif array.dtype.kind == 'b':
-        dtype.code = <uint8_t>kDLBool
+
+    return device
+
+
+# The name of this function is following the framework integration guide of
+# TensorComprehensions.
+cpdef object toDlpack(
```

Comment:
I don't like that it's exposed to Python (it's already done before this PR), but it seems we might need a deprecation cycle before making it `cdef`'d... 😞 (https://github.com/cupy/cupy/issues/8703)